src/boot/efifirmware.c

```diff
@@ -0,0 +1,60 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+
+static const char* efifw_get_fwid(const void *efifwblob) {
+        if ((uintptr_t) efifwblob % alignof(EfiFwHeader) != 0)
+                return NULL;
+
+        const EfiFwHeader *fw_header = ASSERT_PTR(efifwblob);
+
+        if (be32toh(fw_header->magic) != UINT32_C(FWHEADERMAGIC))
+                return NULL;
+
+        uint32_t fw_size = be32toh(fw_header->total_size);
+        uint32_t fw_metadata_off = be32toh(fw_header->fw_metadata_offset);
+        uint32_t fw_metadata_len = be32toh(fw_header->fw_metadata_len);
+        uint32_t fwid_off = be32toh(fw_header->fwid_offset);
+        uint32_t fwid_len = be32toh(fw_header->fwid_len);
+        uint32_t payload_off = be32toh(fw_header->payload_offset);
+        uint32_t payload_len = be32toh(fw_header->payload_len);
+        uint32_t end;
+
+        if (PTR_TO_SIZE(efifwblob) > SIZE_MAX - fw_size)
+                return NULL;
+
+        if (!ADD_SAFE(&end, fw_metadata_off, fw_metadata_len) || end > fw_size)
+                return NULL;
+
+        if (!ADD_SAFE(&end, fwid_off, fwid_len) || end > fw_size)
+                return NULL;
+
+        if (!ADD_SAFE(&end, payload_off, payload_len) || end > fw_size)
+                return NULL;
```

Comment:
i think everything up to here should probably be split out into a function

```c
static const EfiFwHeader *efifw_validate_header(void *blob) { …
```

that simply returns NULL if the header is borked